src/test/java/bio/terra/service/dataset/DatasetBucketDaoTest.java

```diff
@@ -0,0 +1,246 @@
+
+import static org.hamcrest.Matchers.equalTo;
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertFalse;
+import static org.junit.Assert.assertNull;
+import static org.junit.Assert.assertThat;
+import static org.junit.Assert.assertTrue;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+@RunWith(SpringRunner.class)
+@SpringBootTest
+@AutoConfigureMockMvc
+@Category(Unit.class)
+public class DatasetBucketDaoTest {
+
+    private static final Logger logger = LoggerFactory.getLogger(DatasetBucketDaoTest.class);
+
+    @Autowired
+    private JsonLoader jsonLoader;
+
+    @Autowired
+    private DatasetDao datasetDao;
+
+    @Autowired
+    private DatasetBucketDao datasetBucketDao;
+
+    @Autowired
+    private ProfileDao profileDao;
+
+    @Autowired
+    private GoogleResourceDao resourceDao;
+
+    @Autowired
+    private NamedParameterJdbcTemplate jdbcTemplate;
+
+    @Autowired
+    private ResourceService resourceService;
+
+    private BillingProfileModel billingProfile;
+    private GoogleBucketResource bucketForFile;
+    private UUID projectId;
+    private Dataset dataset;
+
+    private UUID datasetId;
+    private UUID bucketResourceId;
+
+    private UUID createDataset(DatasetRequestModel datasetRequest, String newName) throws Exception {
+        datasetRequest.name(newName).defaultProfileId(billingProfile.getId());
+        dataset = DatasetUtils.convertRequestWithGeneratedNames(datasetRequest);
+        dataset.projectResourceId(projectId);
+        String createFlightId = UUID.randomUUID().toString();
+        UUID datasetId = UUID.randomUUID();
+        dataset.id(datasetId);
+        datasetDao.createAndLock(dataset, createFlightId);
+        datasetDao.unlockExclusive(dataset.getId(), createFlightId);
+        return datasetId;
+    }
+
+    private UUID createBucket() throws InterruptedException {
+        String ingestFileFlightId = UUID.randomUUID().toString();
+        bucketForFile =
+            resourceService.getOrCreateBucketForFile(
+                dataset.getName(),
+                billingProfile,
+                ingestFileFlightId);
+        return bucketForFile.getResourceId();
+    }
+
+    private UUID createDataset(String datasetFile) throws Exception  {
+        DatasetRequestModel datasetRequest = jsonLoader.loadObject(datasetFile, DatasetRequestModel.class);
+        return createDataset(datasetRequest, datasetRequest.getName() + UUID.randomUUID().toString());
+    }
+
+    @Before
+    public void setup() {
+        logger.info("-------------------Setup----------------------");
+        BillingProfileRequestModel profileRequest = ProfileFixtures.randomBillingProfileRequest();
+        billingProfile = profileDao.createBillingProfile(profileRequest, "hi@hi.hi");
+
+        GoogleProjectResource projectResource = ResourceFixtures.randomProjectResource(billingProfile);
+        projectId = resourceDao.createProject(projectResource);
```

Comment:
Do we have a billing ID that it's ok for us to use for this test? E.g. every time we run this test, this will create a project...could that expire our quotas?

Reply:
The "createProject" method just inserts a row into the dataset, so it's not actually creating a project.

This is something I've found really confusing about naming conventions.  There are really two layers to our project create/delete: handling the actual google resource versus the metadata about the resource in our database.

Reply:
aaah, ok